timestamp.md

```diff
@@ -0,0 +1,50 @@
+# Timestamping
+
+Time Stamping Authorities (TSAs) defined by [RFC3161](https://tools.ietf.org/html/rfc3161) provide signed timestamp for a signature in order to prove that the signature was generated during the validity period of a certificate.
+
+
+## Scenarios
+
+With TSA, signature can be considered valid even if the signing certificate is expired. This technique is widely used by [Authenticode](https://docs.microsoft.com/windows/win32/seccrypto/time-stamping-authenticode-signatures) with [SignTool](https://docs.microsoft.com/windows/win32/seccrypto/signtool), [NuGet](https://docs.microsoft.com/nuget/create-packages/sign-a-package), [Adobe Acrobat](https://helpx.adobe.com/acrobat/using/certificate-based-signatures.html#add_a_timestamp_to_certificate_based_signatures), and many other industrial products.
+
+In the world of artifacts, including container images, scenarios are
+- Developers sign their artifacts or images with certificates.
+    - The certificates MAY have a configurable expiry time.
+- Developers publish their artifacts or images and at a later date stop maintaining them.
+    - Content consumers SHOULD be able to verify signatures until they expire and use the artifacts or images.
+- Attackers with compromised keys try to sign artifacts or images with timestamps before the key compromise event.
+
+
+## Discussions
+
+There are many public TSA servers [available](https://gist.github.com/Manouchehri/fd754e402d98430243455713efada710) on the Internet. The advantages of public timestamp servers are obvious that they are **public** and **free**.
+
+However, those public TSAs also come with disadvantages:
+- Require the Internet access for signing.
+    - It is not really a disadvantage since public TSAs are online services. Devices in the air-gapped environment SHALL access the Internet for timestamp signing services.
+- Out of the control of the signer.
+    - External dependency.
+    - Availability is not assured. No SLA on signing.
+        - Not all TSAs are available in all regions.
+        - Some regions may have high latency.
+    - The certificates of TSAs MAY be revoked at any time without notices.
+        - [The removal of trust of VeriSign broke .NET 5+ NuGet.](https://github.com/dotnet/announcements/issues/180) Thus Microsoft has to disable the package verification with a new release to unblock customers.
+
+Implications of not using a signed timestamp for a signature:
+- In the absence of additional timestamp signature, the signature is only considered valid till key expiry. This may limit the use of short lived keys.
+- In case of key compromise it's not possible to revoke signatures from a point in time where the time of compromise is known, as an attacker can create signatures with any signature time (by changing local time).
+
+
+## Requirements and Recommendation

```

Comment:
I'm not sure whether this is right place or not(or they have been defined somewhere else) but we should capture requirements around signing and verification:
* During signing, Artifact producer MUST be able to use any TSA(public or private) for timestamping. provides signer agility in case a TSA is down. 
* During signature verification, Artifact consumer MUST be able to enforce presence of timestamp signature.
* During signature verification, Artifact consumer MUST be able to define set of trusted TSAs.

Reply:
>     * During signing, Artifact producer MUST be able to use any TSA(public or private) for timestamping. provides signer agility in case a TSA is down.

I would want to limit that to TSA servers identified by a client side policy. That would allow for disconnected users to point to their local replica, and it would prevent a malicious signer from picking a TSA they control. Producers should pick from a list of TSA servers that they have previously communicated to their consumers.

Reply:
> I would want to limit that to TSA servers identified by a client side policy. 

Does client side policy refers to configurable restrictions inside the signing CLI? 

> Producers should pick from a list of TSA servers that they have previously communicated to their consumers.

Does 3rd requirement `During signature verification, Artifact consumer MUST be able to define set of trusted TSAs.` fits the bill? The idea is that artifact consumer will define the set of trusted TSA's which can be list of TSAs that artifact producer has previously communicated to artifact consumer.

Reply:
> > I would want to limit that to TSA servers identified by a client side policy.
> 
> Does client side policy refers to configurable restrictions inside the signing CLI?

I should have specified, the consumer's policy. I see this as some additional metadata included with the organization's root certificate that will be distributed through an external channel. So the consumer would load the trust store with the cert, and the policy with the list of TSA's to use for that organzation.

> > Producers should pick from a list of TSA servers that they have previously communicated to their consumers.
> 
> Does 3rd requirement `During signature verification, Artifact consumer MUST be able to define set of trusted TSAs.` fits the bill? The idea is that artifact consumer will define the set of trusted TSA's which can be list of TSAs that artifact producer has previously communicated to artifact consumer.

I think it does, but the first bullet still reads with a red flag for me. I'm envisioning an attacker using "any TSA" to include a TSA the attacker controls, allowing them to use an expired key with a malicious TSA to bypass the timestamp checks.